### PR TITLE
perf: efficient implementation of `neighboring`

### LINKED ISSUE #283: Implement the `neighboring` without `exchange` preventing the write of a useless state
Currently, the neighboring operator is implemented in terms of `exchange`. This leads to a useless write into the state that will never be used.
Add the `neighboring` method inside the `Aggregate` interface implemented efficiently.

### PR DESCRIPTION
**Pull request by nicolasfara**
This PR closes #283

### CODE REVIEW

File: dsl/src/commonMain/kotlin/it/unibo/collektive/aggregate/api/Aggregate.kt [hunk 1]
```diff
@@ -63,6 +63,25 @@ interface Aggregate<ID : Any> {
...
+     * val field = neighboring(0)
+     * ```
+     *
+     * The field returned has as local value the value passed as input (0 in this example).
+     *
+     * ```kotlin
+     * val field = neighboring({ 2 * 2 })
+     * ```
+     *
+     * In this case, the field returned has the result of the computation as local value.
```
Comment (DanySK) [inline dsl/src/commonMain/kotlin/it/unibo/collektive/aggregate/api/Aggregate.kt:81]: In this case a Field of `() -> Int` is returned, right?

File: dsl/src/commonTest/kotlin/it/unibo/collektive/aggregate/NeighboringTest.kt [hunk 2]
```diff
@@ -29,6 +29,13 @@ class NeighboringTest : StringSpec({
         }
     }
 
+    "Neighboring without messages with efficient neighboring implementation" {
```
Comment (DanySK) [inline dsl/src/commonTest/kotlin/it/unibo/collektive/aggregate/NeighboringTest.kt:32]: It does not really communicate any intent...

File: dsl/src/commonTest/kotlin/it/unibo/collektive/aggregate/NeighboringTest.kt [hunk 3]
```diff
@@ -56,6 +63,33 @@ class NeighboringTest : StringSpec({
         }
     }
 
+    "Neighboring with three aligned devices with efficient neighboring implementation" {
```
Comment (DanySK) [inline dsl/src/commonTest/kotlin/it/unibo/collektive/aggregate/NeighboringTest.kt:66]: optimized neighboring works across three aligned devices

File: dsl/src/commonTest/kotlin/it/unibo/collektive/aggregate/NeighboringTest.kt [hunk 4]
```diff
@@ -81,4 +115,30 @@ class NeighboringTest : StringSpec({
             field.toMap() shouldHaveSize 1
         }
     }
+
+    "Neighboring with two not aligned devices with efficient neighboring implementation" {
```
Comment (DanySK) [inline dsl/src/commonTest/kotlin/it/unibo/collektive/aggregate/NeighboringTest.kt:119]: non-aligned devices do not communicate

**Review by DanySK — CHANGES_REQUESTED** [review]
minimal comments about docs and tests, the implementation is perfect

File: dsl/src/commonTest/kotlin/it/unibo/collektive/aggregate/NeighboringTest.kt (hunk 2 above)
Comment (nicolasfara) [reply dsl/src/commonTest/kotlin/it/unibo/collektive/aggregate/NeighboringTest.kt]: You are right... I have prepended the string to differentiate the test case but it makes no sense.
I'll change all the test cases in this file to be more explicit.

File: dsl/src/commonMain/kotlin/it/unibo/collektive/aggregate/api/Aggregate.kt (hunk 1 above)
Comment (DanySK) [reply dsl/src/commonMain/kotlin/it/unibo/collektive/aggregate/api/Aggregate.kt:81]: Then the documentation sounds incorrect, the field returned has the computation (and not the result of the computation) as result.

**Review by angelacorte — APPROVED** [review]

File: dsl/src/commonMain/kotlin/it/unibo/collektive/aggregate/api/Aggregate.kt [hunk 5]
```diff
@@ -63,6 +63,25 @@ interface Aggregate<ID : Any> {
...
+     * val field = neighboring(0)
+     * ```
+     *
+     * The field returned has as local value the value passed as input (0 in this example).
+     *
+     * ```kotlin
+     * val field = neighboring({ 2 * 2 })
+     * ```
+     *
+     * In this case, the field returned has the computation as a result.
```
Comment (DanySK) [inline dsl/src/commonMain/kotlin/it/unibo/collektive/aggregate/api/Aggregate.kt:81]:
```suggestion
     * In this case, the field returned has the computation as a result, in form of a field of functions with type `() -> Int`.
```

File: dsl/src/commonTest/kotlin/it/unibo/collektive/aggregate/NeighboringTest.kt [hunk 6]
```diff
@@ -22,14 +22,21 @@ class NeighboringTest : StringSpec({
     val double: (Int) -> Int = { it * 2 }
     val add: (Int) -> Int = { it + 1 }
 
-    "Neighboring without messages" {
+    "Neighbouring must produce a field with the local value when no neighbours are present" {
```
Comment (DanySK) [inline dsl/src/commonTest/kotlin/it/unibo/collektive/aggregate/NeighboringTest.kt:25]:
```suggestion
    "Neighboring must produce a field with the local value when no neighbours are present" {
```

File: dsl/src/commonTest/kotlin/it/unibo/collektive/aggregate/NeighboringTest.kt [hunk 7]
```diff
@@ -22,14 +22,21 @@ class NeighboringTest : StringSpec({
...
         aggregate(id0) {
             val field = neighboringViaExchange(initV1)
             field.toMap() shouldContainValue initV1
         }
     }
 
-    "Neighboring with three aligned devices" {
+    "Optimized neighbouring must produce a field with the local value when no neighbours are present" {
```
Comment (DanySK) [inline dsl/src/commonTest/kotlin/it/unibo/collektive/aggregate/NeighboringTest.kt:32]:
```suggestion
    "Optimized neighboring must produce a field with the local value when no neighbours are present" {
```

**Review by DanySK — CHANGES_REQUESTED** [review]

**Review by DanySK — APPROVED** [review]

**Pull request merged**